engine/access/access_test.go

```diff
@@ -413,6 +415,80 @@ func (suite *Suite) TestGetBlockByIDAndHeight() {
 	})
 }
 
+func (suite *Suite) TestGetExecutionResultByBlockID() {
+	suite.RunTest(func(handler *access.Handler, db *badger.DB, blocks *storage.Blocks, _ *storage.Headers, executionResults *storage.ExecutionResults) {
+
+		// test block1 get by ID
+		nonexistingID := unittest.IdentifierFixture()
+		blockID := unittest.IdentifierFixture()
+
+		er := unittest.ExecutionResultFixture(
+			unittest.WithExecutionResultBlockID(blockID),
+			unittest.WIthServiceEvents(2))
+
+		require.NoError(suite.T(), executionResults.Store(er))
+		require.NoError(suite.T(), executionResults.Index(blockID, er.ID()))
+
+		assertResp := func(resp *accessproto.ExecutionResultForBlockIDResponse, err error, executionResult *flow.ExecutionResult) {
+			require.NoError(suite.T(), err)
+			require.NotNil(suite.T(), resp)
+			er := *resp.ExecutionResult
+
+			require.Len(suite.T(), er.Chunks, len(executionResult.Chunks))
+			require.Len(suite.T(), er.ServiceEvents, len(executionResult.ServiceEvents))
+
+			assert.Equal(suite.T(), executionResult.BlockID[:], er.BlockId)
+			assert.Equal(suite.T(), executionResult.PreviousResultID[:], er.PreviousResultId)
+
+			for i, chunk := range executionResult.Chunks {
+				assert.Equal(suite.T(), chunk.BlockID[:], er.Chunks[i].BlockId)
+				assert.Equal(suite.T(), chunk.Index, er.Chunks[i].Index)
+				assert.Equal(suite.T(), uint32(chunk.CollectionIndex), er.Chunks[i].CollectionIndex)
+				assert.Equal(suite.T(), chunk.StartState[:], er.Chunks[i].StartState)
+				assert.Equal(suite.T(), chunk.EventCollection[:], er.Chunks[i].EventCollection)
+				assert.Equal(suite.T(), chunk.TotalComputationUsed, er.Chunks[i].TotalComputationUsed)
+				assert.Equal(suite.T(), uint32(chunk.NumberOfTransactions), er.Chunks[i].NumberOfTransactions)
+				assert.Equal(suite.T(), chunk.EndState[:], er.Chunks[i].EndState)
+			}
+
+			for i, serviceEvent := range executionResult.ServiceEvents {
+				assert.Equal(suite.T(), serviceEvent.Type, er.ServiceEvents[i].Type)
+				event := serviceEvent.Event
+
+				marshalledEvent, err := json.Marshal(event)
+				require.NoError(suite.T(), err)
+
+				assert.Equal(suite.T(), marshalledEvent, er.ServiceEvents[i].Payload)
+			}
+		}
+
+		suite.Run("nonexisting block", func() {
+			req := &accessproto.GetExecutionResultForBlockIDRequest{
+				BlockId: nonexistingID[:],
+			}
+
+			resp, err := handler.GetExecutionResultForBlockID(context.Background(), req)
+
+			require.Error(suite.T(), err)
+			require.Nil(suite.T(), resp)
+		})
+
+		suite.Run("some block", func() {
+			// get header by ID
+			req := &accessproto.GetExecutionResultForBlockIDRequest{
+				BlockId: blockID[:],
+			}
+
+			resp, err := handler.GetExecutionResultForBlockID(context.Background(), req)
+
+			require.NoError(suite.T(), err)
+
+			assertResp(resp, err, er)
+		})
+
```

Comment:
Is it possible we could exemplify the new logic by essentially reproducing the `TestResultStoreForceIndexOverridesMapping` at this level of the API? That way the semantics are quite a bit more visible.